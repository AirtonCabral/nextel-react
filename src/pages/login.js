import React from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import { loadPage } from '../actions/a_dom'
import { startConnection } from '../actions/a_auth'
import { getProducts } from '../actions/a_products'
import { signIn } from '../actions/a_user'
import CssBaseline from '@material-ui/core/CssBaseline';
import CircularProgress from '@material-ui/core/CircularProgress';
import orange from '@material-ui/core/colors/orange';
import withStyles from '@material-ui/core/styles/withStyles';
import { clearState } from './../lib/storage';

const styles = theme => ({
  layout: {
    width: 'auto',
    display: 'block', // Fix IE11 issue.
    marginLeft: theme.spacing.unit * 3,
    marginRight: theme.spacing.unit * 3,
    [theme.breakpoints.up(400 + theme.spacing.unit * 3 * 2)]: {
      width: 400,
      marginLeft: 'auto',
      marginRight: 'auto',
    },
  },
  paper: {
    marginTop: theme.spacing.unit * 8,
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    padding: `${theme.spacing.unit * 2}px ${theme.spacing.unit * 3}px ${theme.spacing.unit * 3}px`,
  },
  avatar: {
    margin: theme.spacing.unit,
    backgroundColor: '#FF5722',
  },
  form: {
    width: '100%', // Fix IE11 issue.
    marginTop: theme.spacing.unit,
  },
  submit: {
    marginTop: theme.spacing.unit * 3,
    backgroundColor: '#FF5722'
  },
});


//////////////////////////
//  BASE NAME ////////////
//////////////////////////
const basename_home = 'home';
//////////////////////////

export class Login extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      login: '',
      password: '',
      isProcessing: false,
      buttonColorState: 'primary',
      buttonValueState: 'Entrar',
    };
    // console.log('-->> login builded');
  }

  loginApplication() {

    clearState();

    // let { login, password } = this.state;
    let msisdn = this.props.params.msisdn


    this.setState({ isProcessing: true, buttonValueState: 'Iniciando conexão' });

    // this.props.startConnection('drweb', 'n)CJL^?r4p#rYaG/R8A_', msisdn).then(() => { // login producao
    this.props.startConnection('drweb', 'c62J3rZovtw', msisdn).then(() => { // login dev homolog

      let buttonColorResult = this.props.online ? 'secondary' : this.state.buttonColorState;
      let buttonValueResult = this.props.online ? 'Seja bem vindo, Carregando Produtos...' : errorResultMessage;

      this.setState({
        buttonColorState: buttonColorResult,
        buttonValueState: buttonValueResult
      });

      this.props.getProducts().then(() => {
        this.setState({ buttonValueState: 'Baixando dados do Usuário...' });

        this.props.signIn().then(() => {

          // START PROJECT
          if (this.props.assinantesID !== null && this.props.assinantesID !== undefined) {
            setTimeout(() => {
              this.props.loadPage(basename_home)
            }, 300);
            this.setState({ buttonValueState: 'Tudo ok, iniciando...' });
          }
          else {
            this.updateStateOut('Ups, MSISDN Inválido.');
          }

        }) // end success signIn()
          .catch((error) => {
            this.updateStateOut('Ups, usuário indisponível ou instável. Recarregue a tela.');
          });

      }) // end success getProducts()
        .catch((error) => {
          this.updateStateOut('Ups, produtos insdisponíveis para este número "msisdn".');
        });

    }) // end success startConnection()
      .catch((error) => {
        this.updateStateOut('Ups, erro ao tentar se conectar. Recarregue a tela.');
      });

  }

  updateStateOut(msg) {
    this.setState({
      isProcessing: false,
      buttonValueState: msg
    });
  }

  componentDidMount() {
    // console.log('-->> login mounted');
    return this.loginApplication();
  }

  render() {

    const { classes, params } = this.props;
    const isCentral = 'atendente' in params && params.atendente !== '' ? true : false;
    const loginPageTitle = isCentral ? 'Central Serviços VAS' : 'Nextel Serviços VAS'
    const messageOut = this.state.buttonValueState;

    // const errorResultMessage = 'Error  :(  Recarregue a página.';

    // if (this.props.assinantesID !== null && this.props.assinantesID !== undefined) {
    //   messageOut = 'Logado!'
    // }

    if (this.props.userError !== '') {
      return (<div>
        {this.state.buttonValueState}
      </div>)
    }
    const stageHeight = window.innerHeight
    return (
      <div style={{display: 'flex', justifyContent: 'center', alignItems: 'center', height: stageHeight}}>
        <CircularProgress style={{ color: orange[600] }} />
      </div>
    )
  }
}

Login.propTypes = {
  classes: PropTypes.object.isRequired,
};

const mapStateToProps = state => ({
  online: state.auth.online,
  assinantesID: state.user.assinantesID,
  pageLoaded: state.dom.page,
  params: state.dom.params,
  userError: state.user.error,
})

const mapDispatchToProps = dispatch => bindActionCreators({
  startConnection,
  signIn,
  getProducts,
  loadPage,
}, dispatch)

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(withStyles(styles)(Login))

