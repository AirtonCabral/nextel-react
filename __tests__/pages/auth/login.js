import React from 'react';
import renderer from 'react-test-renderer';
import { Login } from '../../../src/pages/auth/login';


/*
    OBSERVE! Teste de Snapshot só é ÚTIL se o desenvolvedor PERMITIR!
    Snapshots só devem ser atualizados se houve mudança de fato na estrutura visual do componente.

    Procedimento Recomendado:
    1. Antes de escrever código que vai atualizar o visual e portanto o snapshot, rode o respectivo teste para
       verificar que este está passando. Se você fizer a mudança antes, o teste falhará e você não saberá se
       foi o seu código ou algum erro anterior.
    2. Não atualize o snapshot que está falhando se você não modificou o seu respectivo componente. O erro
       só ficará  escondido, levando a potenciais bugs difíceis de encontrar.
    3. Evite de gerar snapshots de componentes monolíticos (e.g. algum roteador que contém a estrutura inteira).
       Este raramente será útil e será atualizado constantemente.
    4. Snapshots também deveriam passar por code review! Pode ser que o snapshot gerado não tenha o conteúdo
       que esteja imaginando.

    Veja dicas e mais em - https://blog.kentcdodds.com/effective-snapshot-testing-e0d1a2c28eca

    - joao
*/

describe('Pagina de login', () => {
    it('Renderiza corretamente quando online', () => {
        const tree = renderer.create(<Login online dispatch={() => {}} />).toJSON()
        expect(tree).toMatchSnapshot()
    })

    it('Renderiza corretamente quando offline', () => {
        const tree = renderer.create(<Login online={false} dispatch={() => {}} />).toJSON()
        expect(tree).toMatchSnapshot()
    })

    it('Relata erro quando campos vazios são enviados', () => {
        const login = renderer.create(<Login dispatch={() => {}} />)
        const instance = login.getInstance()

        // Submit o formulário
        instance.contactSubmit()

        // Inspeção dos erros
        const { email, password } = instance.state.errors
        expect(email).toBe('Entre com um endereço de email válido.')
        expect(password).toBe('Sua senha não pode estar em branco.')

        // Força atualização do render
        // FIXME: Não funciona no momento! Snapshot recebe null. Aguardado respostas no stackoverflow.com
        // Segue: https://stackoverflow.com/questions/50417807/jest-snapshot-is-null-after-react-instance-is-updated
        login.update()
        expect(login.toJSON()).toMatchSnapshot()
    })

    // Finalmente teste de sucesso
    it('Não relata erros com dados corretos', () => {
        const login = renderer.create(<Login email="correto@teste.com" password="12345" dispatch={() => {}} />)
        const instance = login.getInstance()

        instance.handleValidation()

        const { email, password } = instance.state.errors
        expect(email).toBeFalsy()
        expect(password).toBeFalsy()
    })

    // Variável mockeada usada como props nos próximos testes.
    const history = {
        push: jest.fn().mockImplementation(() => {}),
    }

    it('Redireciona para o dashboard no sucesso da autenticação', () => {
        // Precisamos aqui de um dispatch mocked mais sofisticado, já que o signin é uma promessa.
        const dispatch = () => new Promise(resolve => resolve())

        // Note também que estamos colocando um token no props, como se o redux tivesse colocado.
        const login = renderer.create(<Login history={history} email="correto@teste.com" password="12345" dispatch={dispatch} token={'token'} />)
        const instance = login.getInstance()

        instance.contactSubmit().then(() => {
            expect(history.push).toBeCalledWith('/dashboard')
        })
    })

    it('Não redireciona no erro da autenticação e reporta o erro', () => {
        history.push.mockReset()
        // Precisamos aqui de um dispatch mocked mais sofisticado, já que o signin é uma promessa.
        const dispatch = () => new Promise(resolve => resolve())

        // Note também que estamos colocando um token no props, como se o redux tivesse colocado.
        const login = renderer.create(<Login history={history} email="correto@teste.com" password="12345" dispatch={dispatch} token={undefined} />)
        const instance = login.getInstance()

        instance.contactSubmit().then(() => {
            const { email } = instance.state.errors
            // expect(history.push).not.toBeCalled()
            expect(email).toBe('Usuário ou senha inválidos')
            expect(history.push).not.toBeCalled()
        })
    })
})
