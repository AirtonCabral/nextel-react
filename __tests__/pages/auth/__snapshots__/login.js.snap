// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagina de login Relata erro quando campos vazios são enviados 1`] = `null`;

exports[`Pagina de login Renderiza corretamente quando offline 1`] = `
<div
  id="login"
>
  <div
    className="bg-login"
  />
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        id="wrapper-form"
      >
        <img
          className="icon-login img-responsive"
          src="test-file-stub"
        />
        <h2
          className="title-login"
        >
          Conecte-se
        </h2>
        <p
          className="text-login"
        >
          Se ainda não tem uma conta, registre-se aqui. 
          <br />
          Se já tem, basta entrar com os dados abaixo.
        </p>
        <form>
          <div
            className="form-group form-group-lg has-feedback"
          >
            <input
              className="form-control"
              disabled={true}
              id="formBasicEmail"
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Email"
              type="text"
              value={undefined}
            />
          </div>
          <div
            className="form-group form-group-lg has-feedback"
          >
            <input
              className="form-control"
              disabled={true}
              id="formBasicPassword"
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Senha"
              type="password"
              value={undefined}
            />
          </div>
          <button
            className="w-100 button-login btn btn-lg btn-default"
            disabled={true}
            type="button"
          >
            SERVIÇO INDISPONÍVEL
          </button>
        </form>
        <a
          className="forgot-password"
          href=""
        >
          Esqueceu sua senha?
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pagina de login Renderiza corretamente quando online 1`] = `
<div
  id="login"
>
  <div
    className="bg-login"
  />
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        id="wrapper-form"
      >
        <img
          className="icon-login img-responsive"
          src="test-file-stub"
        />
        <h2
          className="title-login"
        >
          Conecte-se
        </h2>
        <p
          className="text-login"
        >
          Se ainda não tem uma conta, registre-se aqui. 
          <br />
          Se já tem, basta entrar com os dados abaixo.
        </p>
        <form>
          <div
            className="form-group form-group-lg has-feedback"
          >
            <input
              className="form-control"
              disabled={false}
              id="formBasicEmail"
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Email"
              type="text"
              value={undefined}
            />
          </div>
          <div
            className="form-group form-group-lg has-feedback"
          >
            <input
              className="form-control"
              disabled={false}
              id="formBasicPassword"
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Senha"
              type="password"
              value={undefined}
            />
          </div>
          <button
            className="w-100 button-login btn btn-lg btn-default"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            ENTRAR
          </button>
        </form>
        <a
          className="forgot-password"
          href=""
        >
          Esqueceu sua senha?
        </a>
      </div>
    </div>
  </div>
</div>
`;
